### PR TITLE
refactor: walk the graph using a stream

### LINKED ISSUE #3170: Ensure that Forest and Lotus snapshots match byte-for-byte
**Issue summary**

While restoring the `--tipset` flag for snapshot export in PR #3027, testing of the restored functionality revealed that there appears to be a regression that results in the Forest and Lotus snapshots no longer matching byte-for-byte. This was previously corrected in PR #2540, but the current mismatch may not be related to the `walk_snapshot` function. Investigate and ensure that both clients' snapshots match byte-for-byte.

**Other information and links**

May be fixed in #3173

### PR DESCRIPTION
**Pull request by lemmih**
## Summary of changes

Changes introduced in this pull request:

- Replaced recursive async functions with a stream of all unique nodes in the chain.
- Get rid of the estimated progress when exporting a snapshot. Instead, the size of the snapshot is shown to indicate progress.
- Move handling of temporary files from the daemon into the client.
- Error handling will be improved in a follow-up PR.
- Added CI check to ensure we are aligned with Lotus.

## Reference issue to close (if applicable)

Closes #3170

This PR brings us back in sync with Lotus. Our snapshots are now byte-for-byte identical again.

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by ruseinov** [comment]
As an added bonus we could also resolve this: https://github.com/ChainSafe/forest/pull/3169#discussion_r1261372193

**Review by LesnyRumcajs — APPROVED** [review]

File: src/ipld/util.rs [hunk 1]
```diff
@@ -194,3 +193,146 @@ fn should_save_block_to_snapshot(cid: &Cid) -> bool {
...
+                        let result = this.db.get(&cid);
+
+                        let result = result.and_then(|val| {
+                            let block = val.ok_or(anyhow::anyhow!("missing key"))?;
+                            Ok((cid, block))
+                        });
+
+                        return Poll::Ready(Some(result));
+                    }
+                    Iterate(Ipld::Null) => {}
```
Comment (creativcoder) [inline src/ipld/util.rs:266]: can do `Iterate(Ipld::Null) | Iterate(Ipld::Bool(_))) | ... => {}`

File: src/ipld/util.rs [hunk 2]
```diff
@@ -194,3 +193,146 @@ fn should_save_block_to_snapshot(cid: &Cid) -> bool {
...
+    }
+}
+
+/// Initializes a stream of blocks.
+///
+/// # Arguments
+///
+/// * `db` - A database that implements [`Blockstore`] interface.
+/// * `tipset_iter` - An iterator of [`Tipset`], descending order `$child -> $parent`.
+/// * `stateroot_limit` - An epoch that signifies how far back we need to inspect tipsets.
```
Comment (creativcoder) [inline src/ipld/util.rs:228]: rm `.`?

File: src/ipld/util.rs [hunk 3]
```diff
@@ -194,3 +193,146 @@ fn should_save_block_to_snapshot(cid: &Cid) -> bool {
...
+                        .for_each(|elt| this.dfs.push_front(Iterate(elt))),
+                    Iterate(Ipld::Map(map)) => map
+                        .into_values()
+                        .rev()
+                        .for_each(|elt| this.dfs.push_front(Iterate(elt))),
+                    // The link traversal implementation assumes there are three types of encoding:
+                    // 1. DAG_CBOR: needs to be reachable, so we add it to the queue and load.
+                    // 2. IPLD_RAW: WASM blocks, for example. Need to be loaded, but not traversed.
+                    // 3. _: ignore all other links
+                    Iterate(Ipld::Link(cid)) => {
```
Comment (creativcoder) [inline src/ipld/util.rs:284]: can do:  
```
Iterate(Ipld::Link(cid)) if should_save_block_to_snapshot(cid) && this.seen.insert(cid)  => {

}
```

File: src/ipld/util.rs (hunk 1 above)
Comment (lemmih) [reply src/ipld/util.rs:266]: Pushing to follow-up PR.

File: src/chain/mod.rs [hunk 4]
```diff
@@ -2,5 +2,60 @@
...
+use futures::{io::BufWriter, AsyncWrite};
+use futures_util::future::Either;
+use futures_util::AsyncWriteExt;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_car::CarHeader;
+use std::sync::Arc;
 
 pub use self::{store::*, weight::*};
+
+pub async fn export<W, D>(
```
Comment (sudo-shashank) [inline src/chain/mod.rs:20]: we can add a short description here

File: src/cli/subcommands/archive_cmd.rs [hunk 5]
```diff
@@ -134,43 +119,45 @@ fn build_output_path(chain: String, epoch: ChainEpoch, output_path: PathBuf) ->
...
+    let ts = index.load_tipset(&TipsetKeys::new(store.roots()))?;
 
-    let tmp_chain_dir = TempDir::new()?;
+    let genesis = ts.genesis(&store)?;
+    let network = if genesis.cid() == &*calibnet::GENESIS_CID {
+        NetworkChain::Calibnet
+    } else if genesis.cid() == &*mainnet::GENESIS_CID {
+        NetworkChain::Mainnet
+    } else {
+        NetworkChain::Devnet("devnet".to_string())
```
Comment (sudo-shashank) [inline src/cli/subcommands/archive_cmd.rs:141]: can we use `devnet::GENESIS_CID` here?

**Review by sudo-shashank — APPROVED** [review]

**Pull request merged**